File: majority_vote/cxx/test_majority_vote.cxx
#include <algorithm>
#include <cassert>
#include <cstdlib>
#include <vector>
#include "majority_vote.cxx"

void test_majority_vote()
{
	const int SIZE = 500;
	for(int count = 0; count < SIZE; count++)
	{
		std::vector<int> nums(SIZE, 0);
		for(int i = count; i < nums.size(); i++)
			nums[i] = 1 + rand() % 2;
		std::vector<int> uniq(SIZE);
		std::sort(nums.begin(), nums.end());
		auto it = std::unique_copy(nums.begin(), nums.end(), uniq.begin());
		uniq.resize(std::distance(uniq.begin(), it));
		int expect = -1;
		for(const int& num: uniq)
			if(std::count(nums.begin(), nums.end(), num) > nums.size() / 2)
				expect = num;
		std::random_shuffle(nums.begin(), nums.end());
		assert(majority_vote(nums) == expect && "test majority vote");
	}
}

void test_majority_vote_general()
{
	const int SIZE = 500;
	for(int count = 0; count < SIZE; count++)
	{
		std::vector<int> nums(SIZE, 0);
		int k = 2 + rand() % 8;
		int range = 1 + rand() % 10;
		for(int i = count; i < nums.size(); i++)
			nums[i] = rand() % range;
		std::vector<int> uniq(SIZE);
		std::sort(nums.begin(), nums.end());
		auto it = std::unique_copy(nums.begin(), nums.end(), uniq.begin());
		uniq.resize(std::distance(uniq.begin(), it));
		std::vector<int> expect;
		for(const int& num: uniq)
			if(std::count(nums.begin(), nums.end(), num) > nums.size() / k)
				expect.push_back(num);
		std::random_shuffle(nums.begin(), nums.end());
		std::vector<int> result = majority_vote_general(nums, k);
		assert(std::is_permutation(result.begin(), result.end(), 
					expect.begin()) && "test majority of 1/k votes");
	}
}

void test_majority_vote_edge_case()
{
	const int SIZE = 500;
	for(int count = 0; count < SIZE; count++)
	{
		std::vector<int> nums(SIZE, 0);
		int k = 1 + rand() % SIZE;
		int range = 1 + rand() % 10;
		for(int i = count; i < nums.size(); i++)
			nums[i] = rand() % range;
		std::vector<int> uniq(SIZE);
		std::sort(nums.begin(), nums.end());
		auto it = std::unique_copy(nums.begin(), nums.end(), uniq.begin());
		uniq.resize(std::distance(uniq.begin(), it));
		std::vector<int> expect;
		for(const int& num: uniq)
			if(std::count(nums.begin(), nums.end(), num) > nums.size() / k)
				expect.push_back(num);
		std::random_shuffle(nums.begin(), nums.end());
		std::vector<int> result = majority_vote_general(nums, k);
		assert(std::is_permutation(result.begin(), result.end(), 
					expect.begin()) && "test majority of 1/k votes");
	}
}

int main()
{
	test_majority_vote();
	test_majority_vote_general();
	return 0;
}

File: majority_vote/py3/TestMajorityVote.py
import unittest
import random
from majority_vote import majority_vote, majority_vote_general

class TestMajorityVote(unittest.TestCase):
    def testMajorityVote(self):
        size = random.randrange(100, 500)
        for count in range(size):
            array = [0] * count
            array += [random.randrange(1, 3) for _ in range(size - count)]
            elect = None
            for num in set(array):
                if array.count(num) > len(array) / 2:
                    elect = num
            random.shuffle(array)
            self.assertEqual(majority_vote(array), elect)

    def testMajorityVoteGeneral(self):
        size = random.randrange(100, 500)
        for i in range(100):
            k = random.randrange(2, 10)
            value = random.randrange(1, 10)
            array = [random.randrange(value) for _ in range(size)]
            elect = []
            for num in set(array):
                if array.count(num) > len(array) / k:
                    elect.append(num)
            self.assertEqual(set(elect), set(majority_vote_general(array, k)))

    def testMajorityVoteEdgeCase(self):
        size = random.randrange(100, 500)
        for i in range(100):
            k = random.randrange(1, int(size * 1.5))
            value = random.randrange(1, 10)
            array = [random.randrange(value) for _ in range(size)]
            elect = []
            for num in set(array):
                if array.count(num) > len(array) / k:
                    elect.append(num)
            self.assertEqual(set(elect), set(majority_vote_general(array, k)))

if __name__ == "__main__":
    unittest.main()
